### ISSUE TITLE
Suggestion: abbreviate stack traces by default

### ISSUE THREAD

**Issue by BioTurboNick** [JuliaLang/julia#40138]
I'd like to suggest abbreviating the display of stack traces in errors by default, with the ability to display the full trace immediately after if you wish to see it.

As an example, this stack trace from when `BenchmarkTools.@btime` errors:

```
ERROR: UndefVarError: k not defined
Stacktrace:
 [1] var"##core#271"()
   @ Main ~\.julia\packages\BenchmarkTools\Kt6kv\src\execution.jl:371
 [2] var"##sample#272"(__params::BenchmarkTools.Parameters)
   @ Main ~\.julia\packages\BenchmarkTools\Kt6kv\src\execution.jl:377
 [3] _run(b::BenchmarkTools.Benchmark{Symbol("##benchmark#270")}, p::BenchmarkTools.Parameters; verbose::Bool, pad::String, kwargs::Base.Iterators.Pairs{Symbol, Integer, NTuple{4, Symbol}, NamedTuple{(:samples, :evals, :gctrial, :gcsample), Tuple{Int64, Int64, Bool, Bool}}})
   @ Main ~\.julia\packages\BenchmarkTools\Kt6kv\src\execution.jl:405
 [4] #invokelatest#2
   @ .\essentials.jl:710 [inlined]
 [5] #run_result#39
   @ ~\.julia\packages\BenchmarkTools\Kt6kv\src\execution.jl:32 [inlined]
 [6] run(b::BenchmarkTools.Benchmark{Symbol("##benchmark#270")}, p::BenchmarkTools.Parameters; progressid::Nothing, nleaves::Float64, ndone::Float64, kwargs::Base.Iterators.Pairs{Symbol, Integer, NTuple{5, Symbol}, NamedTuple{(:verbose, :samples, :evals, :gctrial, :gcsample), Tuple{Bool, Int64, Int64, Bool, Bool}}})
   @ BenchmarkTools ~\.julia\packages\BenchmarkTools\Kt6kv\src\execution.jl:94
 [7] #warmup#47
   @ ~\.julia\packages\BenchmarkTools\Kt6kv\src\execution.jl:141 [inlined]
 [8] warmup(item::BenchmarkTools.Benchmark{Symbol("##benchmark#270")})
   @ BenchmarkTools ~\.julia\packages\BenchmarkTools\Kt6kv\src\execution.jl:141
 [9] top-level scope
   @ ~\.julia\packages\BenchmarkTools\Kt6kv\src\execution.jl:481
```

An improved experience might be something like:
```
ERROR: UndefVarError: k not defined
in var"##core#271"()
   @ Main ~\.julia\packages\BenchmarkTools\Kt6kv\src\execution.jl:371
Use `stacktrace(errs)` to retrieve the full stack trace.
```

Related to #36517, where @vtjnash mentioned the possibility of a `errs` variable that stores an error similar to `ans` storing the last result.

**Comment by KristofferC** [comment]
> An improved experience might be something like:

So always take the first? How about:

```
julia> "foo"[4]
ERROR: BoundsError: attempt to access 3-codeunit String at index [4]
Stacktrace:
 [1] checkbounds
   @ ./strings/basic.jl:216 [inlined]
 [2] codeunit
   @ ./strings/string.jl:102 [inlined]
 [3] getindex(s::String, i::Int64)
   @ Base ./strings/string.jl:223
 [4] top-level scope
   @ REPL[2]:1
```

Would only showing the first entry there really be better?

**Comment by BioTurboNick** [comment]
I think so? The error message itself is pretty clear, the stack trace seems dispensable?

**Comment by BioTurboNick** [comment]
Or you could show the first non-Base/stdlib frame

**Comment by KristofferC** [comment]
> I think so? The error message itself is pretty clear, the stack trace seems dispensable?

But how would you know where in your code the problem originated from?

**Comment by JeffBezanson** [comment]
This has come up a couple times in the past and I generally resist hiding information. To me, the tradeoff is not good since at best this just avoids a bit of visual noise, but at worst it makes debugging harder. We will start to see bug reports without the whole stack trace. Obviously many people will request the full trace before filing an issue, but it's one more step in the way of getting the information we need. To be a bit hyperbolic, it's like complaining that a smoke detector is too loud.

**Comment by BioTurboNick** [comment]
I totally understand that view.

To a novice user though, extraneous information can be daunting and actively prevent them from understanding where the problem is, potentially leading to unnecessary support requests or frustration.

I was partly inspired by the complaint posted on Slack from the JuliaOptics guy, but I've seen that complaint before and it was a bit of a barrier for me getting started.

Readily available, discoverable, but not shown by default strikes a good balance there IMO. I made a similar argument in favor of low-cost notice of deprecation warnings, against them being hidden entirely.

Take this example:
```
a() = "foo"[4]
b() = a()
c() = b()
d() = c()
e() = d()
e()

ERROR: BoundsError: attempt to access 3-codeunit String at index [4]
Stacktrace:
 [1] checkbounds
   @ .\strings\basic.jl:216 [inlined]
 [2] codeunit
   @ .\strings\string.jl:102 [inlined]
 [3] getindex
   @ .\strings\string.jl:223 [inlined]
 [4] a
   @ .\REPL[6]:1 [inlined]
 [5] b
   @ .\REPL[7]:1 [inlined]
 [6] c
   @ .\REPL[8]:1 [inlined]
 [7] d
   @ .\REPL[9]:1 [inlined]
 [8] e()
   @ Main .\REPL[10]:1
 [9] top-level scope
   @ REPL[11]:1

vs.

ERROR: BoundsError: attempt to access 3-codeunit String at index [4]
in a
   @ .\REPL[6]:1 [inlined]
 Use `stacktrace(errs)` to retrieve the full stack trace.
```

If a stack trace can be filtered for Base/stdlib methods and show the top one, it would be a lot more clear where those more simple errors originate.

Maybe there are good reasons not to do it right now, but it might be useful as Julia matures and grows.

**Comment by david-macmahon** [comment]
Is there some way to override how the REPL displays exceptions?  I'm imagining something like a user-supplied `show` method in `startup.jl` to handle how Exceptions are displayed, but I don't know how feasible something like that is.

**Comment by BioTurboNick** [comment]
Just learned about `isinteractive`. Is it the case that `isinteractive` will be false e.g. in CI runs, where if a feature like this was adopted, the whole stack could be dumped normally in that context?

**Comment by StefanKarpinski** [comment]
I want to add a voice from the "core dev" side in support of abbreviated stack traces. I think this is a major usability issue for the language. Two-language solutions have an advantage here: the boundary between the high-level (e.g. Python) and low-level language (e.g. C) is a natural stopping point for errors — if something is wrong about how the user is calling a function, it will get caught before calling into C or not at all; if something goes wrong inside of the C call, the user doesn't get a C stack trace, they see an error raised by the C function. Imagine if Python users were subjected to C stack traces deep into the CPython runtime? That's basically what we're currently subjecting Julia users to. Yes, the stack is all Julia calls, but if something goes deep inside of stdlib, does the typical user they care about a dozen internal function calls between the public stdlib function they called and where they error actually occurred? They do not. Printing all that extra information makes it much harder for the typical user to understand what actually went wrong and fix it. Frankly, when I get stack traces, I find it annoyingly difficult to figure out where the actual problem is in the wall of text that we dump.

**Comment by stevengj** [comment]
What if we restricted the stack trace by default to show (a) lines in code defined in Main or any modules in untracked/dev packages, (b) the entry points into all other code, and (c) the topmost stackframe. 

The [example above](https://github.com/JuliaLang/julia/issues/40138#issuecomment-804310686) would become:
```
 [1] request(::Type{ExceptionLayer{ConnectionPoolLayer{StreamLayer{Union{}}}}}, ::URIs.URI, ::Vararg{Any,N} where N; kw::Base.Iterators.Pairs{Symbol,Any,Tuple{Symbol,Symbol,Symbol},NamedTuple{(:iofunction, :reached_redirect_limit, :accept),Tuple{Nothing,Bool,String}}}) at C:\Users\nicho\.julia\packages\HTTP\cxgat\src\ExceptionRequest.jl:22
 [2...6] (omitted) 
 [7] request(::String, ::String, ::Array{Pair{SubString{String},SubString{String}},1}, ::Array{UInt8,1}; headers::Array{Pair{SubString{String},SubString{String}},1}, body::Array{UInt8,1}, query::Nothing, kw::Base.Iterators.Pairs{Symbol,String,Tuple{Symbol},NamedTuple{(:accept,),Tuple{String}}}) at C:\Users\nicho\\.julia\packages\HTTP\cxgat\src\HTTP.jl:315
 [8] top-level scope at REPL[17]:1
```

**Comment by BioTurboNick** [comment]
That's basically what my split-off package does right now if you haven't taken a look yet, @stevengj. My plan is to refine the idea there and then it can be brought back when it seems most issues are addressed.

**Comment by stevengj** [comment]
Also, I see that @vtjnash suggested an alternative heuristic [here](https://github.com/JuliaLang/julia/pull/40537#issuecomment-828122244).   However, I'm not convinced that "crossing API" boundaries is the right heuristic.  e.g. if `DifferentialEquations` calls something in `DiffEqBase` which calls something in `ForwardDiff` which calls something in `ChainRules`, none of that is *my* code so I'm not sure why I should care just because they jumped packages.  And indeed, long chains of dependencies like this are quite normal in Julia.

**Comment by ChrisRackauckas** [comment]
Bump on this discussion. While I understand the argument that we want stacktraces to be complete so that when posts them onto the Discourse we have all of the information to debug from, it turns out that the current strategy of "print everything" actually stops that from occurring in practice. I mean, the stacktraces are so long now that we post pictures of them in chats as jokes.

![image](https://user-images.githubusercontent.com/1814174/139518394-6b658876-10c4-4403-a440-f5b432478d53.png)

Do we really think people are going to copy-paste all of that?

> We will start to see bug reports without the whole stack trace.

We won't start to, we already do see people cut stack traces. What we see is gigantic 1000 line stacktraces that confuse users, which makes them think it's all a bunch of junk. This leads people to shorten it themselves (like in https://discourse.julialang.org/t/incorporating-forcing-functions-in-the-ode-model/70133/5). Making people who don't know how to read stacktraces be the ones to figure out what to delete is the worst of all worlds. I think the majority of heuristics would do better than most user's ideas (and the average Discourse user seems to generally take the suggestion of the OP to just take the top of the stack. That almost never points to the solution so I would heavily oppose that).

**Comment by mcabbott** [comment]
> abbreviate the display of long parameterized type names

What's a good heuristic here? Within the text which is currently light grey, it could (1) just cap the length at 1 screen width or so? (2) cap the number of levels of nesting and print? Still printing all the arguments (thus all the bits of white text at the end of [10] here).

This seems largely orthogonal deciding which frames to print, and perhaps easier?

> which calls something in ForwardDiff which calls something in ChainRules, none of that is my code so I'm not sure why I should care

One reason to nevertheless print things is that the people you ask for help may know about these packages.

> what my split-off package does right now

This isn't linked here, but is https://github.com/BioTurboNick/AbbreviatedStackTraces.jl . This is, if I understand right, primarily about deciding which frames to print.

**Comment by BioTurboNick** [comment]
> This isn't linked here, but is https://github.com/BioTurboNick/AbbreviatedStackTraces.jl . This is, if I understand right, primarily about deciding which frames to print.

Ah thanks for linking my package @mcabbott, and to its related draft PR #40537, which it seems I neglected to link back to this issue. And thanks @ChrisRackauckas for bumping it.

The solution I settled on there was to hide "internal" frames by default, but they would be viewable in full on demand by accessing an `err` global. But that's still open for feedback. There's also a mode for stripping out type information altogether so each frame fits on one line; which may be too much reduction.

But I hadn't done anything about type information otherwise because I haven't had direct experience of the problem, enough to have an idea of how to handle it. That screenshot *is* pretty crazy.

This might not be exactly what my package would produce (I haven't tested it with nested task exceptions, and it could probably be abbreviated further than this), but in theory the stack trace in the Discourse thread would be abbreviated to:
```
┌ Warning: Only a single thread available: MCMC chains are not sampled in parallel      
└ @ AbstractMCMC C:\Users\Bharadwaj\.julia\packages\AbstractMCMC\BPJCW\src\sample.jl:291
ERROR: LoadError: TaskFailedException

    nested task error: TaskFailedException       
    Stacktrace:
     [1-4] internal
         @ Base.Threads, AbstractMCMC
       [5] (::AbstractMCMC.var"#31#41"{Bool, Base.Iterators.Pairs{Symbol, UnionAll, Tuple{Symbol}, NamedTuple{(:chain_type,), Tuple{UnionAll}}}, Int64, Int64, Vector{Any}, Vector{UInt64}, Vector{DynamicPPL.Sampler{NUTS{Turing.Core.ForwardDiffAD{40}, (), AdvancedHMC.DiagEuclideanMetric}}}, Vector{DynamicPPL.Model{typeof(fitting_epidemic_wildtype), (:observ_data, :w_forcing, :m_forcing), (), (), Tuple{Matrix{Float64}, Interpolations.BSplineInterpolation{Float64, 1, Vector{Float64}, BSpline{Linear{Throw{OnGrid}}}, Tuple{Base.OneTo{Int64}}}, Interpolations.BSplineInterpolation{Float64, 1, Vector{Float64}, BSpline{Linear{Throw{OnGrid}}}, Tuple{Base.OneTo{Int64}}}}, Tuple{}, DynamicPPL.DefaultContext}}, Vector{Random._GLOBAL_RNG}})()
         @ AbstractMCMC .\task.jl:411

        nested task error: MethodError: no constructors have been defined for Any
        Stacktrace:
           [1-5] internal
               @ Base
             [6] fitting_epidemic_wildtype(__model__::DynamicPPL.Model{typeof(fitting_epidemic_wildtype), (:observ_data, :w_forcing, :m_forcing), (), (), Tuple{Matrix{Float64}, Interpolations.BSplineInterpolation{Float64, 1, Vector{Float64}, BSpline{Linear{Throw{OnGrid}}}, Tuple{Base.OneTo{Int64}}}, Interpolations.BSplineInterpolation{Float64, 1, Vector{Float64}, BSpline{Linear{Throw{OnGrid}}}, Tuple{Base.OneTo{Int64}}}}, Tuple{}, DynamicPPL.DefaultContext}, __varinfo__::DynamicPPL.UntypedVarInfo{DynamicPPL.Metadata{Dict{AbstractPPL.VarName, Int64}, Vector{Distribution}, Vector{AbstractPPL.VarName}, Vector{Real}, Vector{Set{DynamicPPL.Selector}}}, Float64}, __context__::DynamicPPL.SamplingContext{DynamicPPL.SampleFromUniform, DynamicPPL.DefaultContext, Random._GLOBAL_RNG}, observ_data::Matrix{Float64}, w_forcing::Interpolations.BSplineInterpolation{Float64, 1, Vector{Float64}, BSpline{Linear{Throw{OnGrid}}}, Tuple{Base.OneTo{Int64}}}, m_forcing::Interpolations.BSplineInterpolation{Float64, 1, 
Vector{Float64}, BSpline{Linear{Throw{OnGrid}}}, Tuple{Base.OneTo{Int64}}})
               @ Main c:\Users\Bharadwaj\Indian Institute of Science\COVID-19 variant study - General\codes_wildtype_fitting\discourse_NPI_code.jl:100
          [7-20] internal
               @ DynamicPPL, AbstractMCMC, Turing
            [21] (::AbstractMCMC.var"#927#threadsfor_fun#42"{UnitRange{Int64}, Bool, Base.Iterators.Pairs{Symbol, UnionAll, Tuple{Symbol}, NamedTuple{(:chain_type,), Tuple{UnionAll}}}, Int64, Vector{Any}, Vector{UInt64}, Vector{DynamicPPL.Sampler{NUTS{Turing.Core.ForwardDiffAD{40}, (), AdvancedHMC.DiagEuclideanMetric}}}, Vector{DynamicPPL.Model{typeof(fitting_epidemic_wildtype), (:observ_data, :w_forcing, :m_forcing), (), (), Tuple{Matrix{Float64}, Interpolations.BSplineInterpolation{Float64, 1, Vector{Float64}, BSpline{Linear{Throw{OnGrid}}}, Tuple{Base.OneTo{Int64}}}, Interpolations.BSplineInterpolation{Float64, 1, Vector{Float64}, BSpline{Linear{Throw{OnGrid}}}, Tuple{Base.OneTo{Int64}}}}, Tuple{}, DynamicPPL.DefaultContext}}, Vector{Random._GLOBAL_RNG}})()
               @ AbstractMCMC .\threadingconstructs.jl:48
Stacktrace:
 [1-11] internal
      @ Base, AbstractMCMC, ProgressLogging, Turing
   [12] sample(model::DynamicPPL.Model{typeof(fitting_epidemic_wildtype), (:observ_data, :w_forcing, :m_forcing), (), (), Tuple{Matrix{Float64}, Interpolations.BSplineInterpolation{Float64, 1, Vector{Float64}, BSpline{Linear{Throw{OnGrid}}}, Tuple{Base.OneTo{Int64}}}, Interpolations.BSplineInterpolation{Float64, 1, Vector{Float64}, BSpline{Linear{Throw{OnGrid}}}, Tuple{Base.OneTo{Int64}}}}, Tuple{}, DynamicPPL.DefaultContext}, alg::NUTS{Turing.Core.ForwardDiffAD{40}, (), AdvancedHMC.DiagEuclideanMetric}, ensemble::MCMCThreads, N::Int64, n_chains::Int64)
      @ Turing.Inference C:\Users\Bharadwaj\.julia\packages\Turing\uMQmD\src\inference\Inference.jl:189
   [13] top-level scope
      @ c:\Users\Bharadwaj\Indian Institute of Science\COVID-19 variant study - General\codes_wildtype_fitting\discourse_NPI_code.jl:120
in expression starting at c:\Users\Bharadwaj\Indian Institute of Science\COVID-19 variant study - General\codes_wildtype_fitting\discourse_NPI_code.jl:120
```

**Comment by mauro3** [comment]
How about  not printing the type parameters if they exceed, say 20 characters.  That would reduce Chris' stack trace a lot, e.g. frame 10 would be:
```
[10] generateFMU(surr::LPCTNSNSurrogate{...}, sys:ODESystem, u0:Vector{Float64}, etc
```

**Comment by mcabbott** [comment]
Maybe this should be a separate issue from the other sense of abbreviation? Anyway this is my (1) above, which is dead-simple to apply to this function: https://github.com/JuliaLang/julia/blob/master/base/show.jl#L2382-L2393, as in [this gist](https://gist.github.com/mcabbott/e679979d437de5ad48dddc367e02599e).

**Comment by BioTurboNick** [comment]
I should correct myself: My package does currently elide type parameters when there are more than 2. But I don't propose that's the best rule.

It does make sense to talk about that separately.

**Comment by timholy** [comment]
I am not a fan of heuristics. I think the better way to solve this is to create `typefold(T)`, with the name being chosen to be suggestive of [code folding](https://en.wikipedia.org/wiki/Code_folding). It would processes the type to try to get back to something that looks like the signature of the method called. E.g., `foo(x::T, y::T)` where `T` is some complicated, long-printing type gets returned as something that will print as `foo(x::T, y::T) where T = <first "layer" of complicated type>`. (The point being, a literal `T` gets printed twice rather than printing the complicated type twice, and in lieu of printing the entire complicated type we open with just its outermost layer.) In an ideal world, `Union`s like the infamous `StridedArray` get packaged back up as `StridedArray`. `typefold` returns a struct aimed at efficient printing with code-folded display.

Then, we also modify the REPL so that we can interactively fold and unfold portions of the type. https://github.com/JuliaCollections/FoldingTrees.jl already does this but it's in a multiline format, and for type-printing I think we want an inline variant. This should be done in a way that I can use with an arbitrary type that I call `typefold` on.

Finally, we modify stacktraces to use `typefold` on each line. We add another REPL feature, similar to the Ctrl-Q trick, that lets you scroll up and down the stacktrace and expand chosen types.

All this should be done in a manner that allows VS Code to do similar operations but via mouse clicks.

And yes, I agree with @ChrisRackauckas and others that this is a first-order usability issue.

**Comment by vtjnash** [comment]
You mean this infamous `StridedArray`?
```julia
julia> StridedArray
StridedArray (alias for Union{DenseArray{T, N}, Base.ReinterpretArray{T, N, S, A, IsReshaped} where {A<:Union{SubArray{T, N, A, I, true} where {T, N, A<:DenseArray, I<:Union{Tuple{Vararg{Real}}, Tuple{AbstractUnitRange, Vararg{Any}}}}, DenseArray}, IsReshaped, S}, Base.ReshapedArray{T, N, A} where A<:Union{Base.ReinterpretArray{T, N, S, A, IsReshaped} where {T, N, A<:Union{SubArray{T, N, A, I, true} where {T, N, A<:DenseArray, I<:Union{Tuple{Vararg{Real}}, Tuple{AbstractUnitRange, Vararg{Any}}}}, DenseArray}, IsReshaped, S}, SubArray{T, N, A, I, true} where {T, N, A<:DenseArray, I<:Union{Tuple{Vararg{Real}}, Tuple{AbstractUnitRange, Vararg{Any}}}}, DenseArray}, SubArray{T, N, A, I} where {A<:Union{Base.ReinterpretArray{T, N, S, A, IsReshaped} where {T, N, A<:Union{SubArray{T, N, A, I, true} where {T, N, A<:DenseArray, I<:Union{Tuple{Vararg{Real}}, Tuple{AbstractUnitRange, Vararg{Any}}}}, DenseArray}, IsReshaped, S}, Base.ReshapedArray{T, N, A} where {T, N, A<:Union{Base.ReinterpretArray{T, N, S, A, IsReshaped} where {T, N, A<:Union{SubArray{T, N, A, I, true} where {T, N, A<:DenseArray, I<:Union{Tuple{Vararg{Real}}, Tuple{AbstractUnitRange, Vararg{Any}}}}, DenseArray}, IsReshaped, S}, SubArray{T, N, A, I, true} where {T, N, A<:DenseArray, I<:Union{Tuple{Vararg{Real}}, Tuple{AbstractUnitRange, Vararg{Any}}}}, DenseArray}}, DenseArray}, I<:Tuple{Vararg{Union{Int64, AbstractRange{Int64}, Base.AbstractCartesianIndex, Base.ReshapedArray{T, N, A, Tuple{}} where {T, N, A<:AbstractUnitRange}}}}}} where {T, N})

julia> f(::StridedArray) = 1
f (generic function with 1 method)

julia> methods(f)
# 1 method for generic function "f":
[1] f(::StridedArray) in Main at REPL[3]:1
```

**Comment by juliohm** [comment]
Any plans to add more visual elements in stacktraces a la https://github.com/FedeClaudi/Term.jl ?

Maybe adding boxes around the main error message could help?

**Comment by BioTurboNick** [comment]
That is pretty! That could be a separate issue, or demonstration package, possibly? Though if you have an idea that would look particularly great with the abbreviated traces I came up with here: https://github.com/BioTurboNick/AbbreviatedStackTraces.jl I'd welcome an issue or PR.

**Comment by ChrisRackauckas** [comment]
I had a very concrete way to shorten a lot of stack traces without losing any information while also making code easier to maintain, so I opened that in https://github.com/JuliaLang/julia/issues/45687 . It would solve the long stack trace problem for SciML at least.

**Comment by sloede** [comment]
I would like to voice my support for improving the current state of stack traces, and soon. As a real-world example, when using OrdinaryDiffEq.jl for time integration, we regularly get error stacktraces for Trixi.jl that are around 80 KB in size, where two stack frames alone have >25k characters. If I trigger such an error with multithreading enabled (e.g., 32 threads), this is even worse.

I truly appreciate the fantastic job the OrdinaryDiffEq.jl folks do, and I think they are doing everything right in leveraging Julia's capabilities to the full extent. However, I believe the tooling (in this case: stack traces) should keep up with such "extreme" usage of the Julia machinery (specifically, the type parameters).

A one-size-fits-all solution for this issue is likely impossible, but if we were able to start with one of the proposed approaches (either radically shortening type parameter printing or omitting some frames), we would be able to gather experience with this in practice. Of course, there should be a command line flag + possible an environment variable to restore the previous behavior (e.g., to globally turn off stacktrace shortening) and/or an easy option to recreate the full stack trace (like `err` does for https://github.com/BioTurboNick/AbbreviatedStackTraces.jl), but I believe overall it would greatly benefit many users (and developers!) of advanced Julia packages.

**Comment by ChrisRackauckas** [comment]
https://github.com/JuliaLang/julia/issues/45687 is something we want to follow up on.

**Comment by BioTurboNick** [comment]
> A one-size-fits-all solution for this issue is likely impossible, but if we were able to start with one of the proposed approaches (either radically shortening type parameter printing or omitting some frames), we would be able to gather experience with this in practice. Of course, there should be a command line flag + possible an environment variable to restore the previous behavior (e.g., to globally turn off stacktrace shortening) and/or an easy option to recreate the full stack trace (like `err` does for https://github.com/BioTurboNick/AbbreviatedStackTraces.jl), but I believe overall it would greatly benefit many users (and developers!) of advanced Julia packages.

`err` has already been implemented in the Julia REPL. (Though the PR for adding this to the VS Code REPL is waiting for approval still)

**Comment by sloede** [comment]
> `err` has already been implemented in the Julia REPL.

Where can I find this/how can I use this - is this documented in the Julia manual somewhere? It does not solve the issue of the overly large stacktraces though, or does it?

**Comment by BioTurboNick** [comment]
> > `err` has already been implemented in the Julia REPL.
> 
> Where can I find this/how can I use this - is this documented in the Julia manual somewhere? It does not solve the issue of the overly large stacktraces though, or does it?

No, just pointing that out. It was in NEWS for 1.8 but it appears some docs are missing. Should probably be next to `ans` here? https://docs.julialang.org/en/v1/base/base/#ans

**Comment by PallHaraldsson** [comment]
Can we have abbreviated stack traces (by default), and also output the full one elsewhere?

The example Chis had is bad, and worse if people don't know where to cut. The short version could end with full stack trace is at /tmp/something. It's unclear to me if advised to overwrite this file each time, or log-rotate, keep last 3 or whatever, maybe configurable. Or put it into syslong (maybe only if opted into with ENV var)?

Could we even pop up a windows with short, and details button?

**Comment by BioTurboNick** [comment]
I think writing exceptions to a file would be a great idea. Could do a separate issue/PR for that, as that would possibly be useful regardless of whether traces get abbreviated, though it would work well with it. And might get better discussion than here. If nothing else, it could just copy what gets written to the `err` variable, and anything fancier could come later.

Invoking anything with a GUI is probably a no-go in Base though.

I think though the stacktrace in the issue you're referencing (from the Julia process crashing) is a different matter than the traces produced internally in a normal Julia session - different code path, different considerations.

**Comment by PallHaraldsson** [comment]
I wouldn't want to write to a file for non-erroneous situations, why I mentioned SIGTERM also, which is considered normal. Another thing, when I install packages, I often stop the pre-compilation process intentionally, and get huge stacktraces. I'm not really interested in them, even if abbreviated. A separate PR could drop them entirely?

But for real valid stacktraces, I would be very ok with abbrevated (similar to in Python, stopping at the C API), and the file idea I proposed. Would anyone object to that?

I'm thinking could this be considered a privacy issue, or bad for diskspace reasons? The GUI idea was to possibly get out of wasting disk space, log-rotation would also. But it seems bad to abbreviate only and have no way of getting full, since that can be helpful. If current stack-traces go to logs in some situations, and to many generated (never good) then we would lose out with log-rotation, only keeping abbveviated in syslog...

**Comment by BioTurboNick** [comment]
I made a proposal on the PR, in case people have participated on the issue and not the PR: https://github.com/JuliaLang/julia/pull/40537#issuecomment-2194859995

Maybe it is a faster way forward?

**Comment by iago-lito** [comment]
Hello. I have been kindly prompted by @adienes to comment here so as to maybe help this forward.

I would like to showcase two situations which I believe are fairly common and in which I think it is _unambiguous_ which part of the stack trace is "irrelevant".
Only the responsibility to abbreviate falls to _lib developers_, not to julia. So they would not be abbreviated "by default".
Although we'd still need julia to make it possible.

## Situation A: drop all irrevelant frames to the bottom.

Here is what happens today in this situation:
```julia
#-------------------------------------
module Internal
struct WrongInput <: Exception end
f(_) = throw(WrongInput())
f(x::Int) = x
g(x) = f(x)
h(x) = g(x)
export h
end

#-------------------------------------
module Exposed

using ..Internal
struct NiceReport <: Exception end

p(x) = h(x)
q(x) = p(x)
welcome(x) =
    try
        q(x)
    catch e
        e isa Internal.WrongInput || rethrow(e)
        rethrow(NiceReport())
    end
export welcome

Base.showerror(io::IO, ::NiceReport) =
    print(io, "Don't worry it's not that bad just try again <3")

end

#-------------------------------------
# User.
using .Exposed

my_u(x) = welcome(x)
my_v(x) = my_u(x)
my_w(x) = my_v(x)

my_w("hi")
```
```text
ERROR: LoadError: Don't worry it's not that bad just try again <3
Stacktrace:
  [1] f(::String)                <<
    @ Main.Internal ./test.jl:4  <<
  [2] g(x::String)               <<  Garbage (from user perspective).
    @ Main.Internal ./test.jl:6  <<  Useless for debugging: not a bug (in the lib).
  [3] h(x::String)               <<
    @ Main.Internal ./test.jl:7  <<
  [4] p(x::String)               <<
    @ Main.Exposed ./test.jl:17  <<
  [5] q(x::String)               <<
    @ Main.Exposed ./test.jl:18  <<
  [6] welcome(x::String)         <<
    @ Main.Exposed ./test.jl:21  <<
  [7] my_u(x::String)         <<
    @ Main ./test.jl:37       <<
  [8] my_v(x::String)         <<  *USEFUL* (from user perspective).
    @ Main ./test.jl:38       <<
  [9] my_w(x::String)         <<
    @ Main ./test.jl:39       <<
 [10] top-level scope         <<
    @ ./test.jl:41            <<
in expression starting at ./test.jl:41
```

In order to fix this, I suggest that the situation remains the same, but that developer of the `Exposed` module _can_ now opt into trimming frames `[1]` to `[6]` by changing something along:
```diff
- rethrow(NiceReport())
+ throw(NiceReport(); afresh = true)
```
Meaning that all stack frames and exceptions prior to this enclosing `catch` block may be safely dropped as they are guaranteed irrelevant by `Exposed`.

(Notice how the stack remains as deep as can be if the error caught is not of the expected kind, indicating a bug in `Exposed`.)

## Situation B: stitch around irrelevant part.

Here is a somewhat more intricate situation:
```julia
#-------------------------------------
module Internal
struct CallBackError <: Exception
    user_err
end
f(callback::Function) =
    try
        callback()
    catch e
        rethrow(CallBackError(e))
    end
g(c) = f(c)
h(c) = g(c)
export h
end

#-------------------------------------
module Exposed

using ..Internal
struct NiceReport <: Exception
    user_err
end

p(c) = h(c)
q(c) = p(c)
welcome(c) =
    try
        q(c)
    catch e
        e isa Internal.CallBackError || rethrow(e)
        rethrow(NiceReport(e.user_err))
    end
export welcome

function Base.showerror(io::IO, r::NiceReport)
    println(io, "Your callback failed with the following error:")
    showerror(io, r.user_err)
    print(io, "Don't worry it's not that bad just try again <3")
end

end

#-------------------------------------
# User.
using .Exposed

my_u(c) = welcome(c)
my_v(c) = my_u(c)
my_w(c) = my_v(c)

my_x() = :oops + 1
my_y() = my_x()
my_z() = my_y()

my_w(my_z)
```
```text
ERROR: LoadError: Your callback failed with the following error:
MethodError: no method matching +(::Symbol, ::Int64)
The function `+` exists, but no method is defined for this combination of argument types.

Closest candidates are:
  +(::Any, ::Any, !Matched::Any, !Matched::Any...)
   @ Base operators.jl:642
  +(!Matched::Base.CoreLogging.LogLevel, ::Integer)
   @ Base logging./test.jl:134
  +(!Matched::BigInt, ::Union{Int16, Int32, Int64, Int8})
   @ Base gmp.jl:556
  ...
Don't worry it's not that bad just try again <3
Stacktrace:
  [1] my_x()                   <<
    @ Main ./test.jl:52        <<
  [2] my_y()                   <<  *USEFUL* (from user perspective).
    @ Main ./test.jl:53        <<
  [3] my_z()                   <<
    @ Main ./test.jl:54        <<
  [4] f(callback::typeof(my_z))   <<
    @ Main.Internal ./test.jl:8   <<
  [5] g(c::Function)              <<
    @ Main.Internal ./test.jl:12  <<  Garbage (from user perspective).
  [6] h                           <<  Useless for debugging: not a bug (in the lib).
    @ ./test.jl:13 [inlined]      <<
  [7] p                           <<
    @ ./test.jl:25 [inlined]      <<
  [8] q                           <<
    @ ./test.jl:26 [inlined]      <<
  [9] welcome(c::Function)        <<
    @ Main.Exposed ./test.jl:29   <<
 [10] my_u                     <<
    @ ./test.jl:48 [inlined]   <<
 [11] my_v                     <<  *USEFUL* (from user perspective).
    @ ./test.jl:49 [inlined]   <<
 [12] my_w(c::Function)        <<
    @ Main ./test.jl:50        <<
 [13] top-level scope          <<
    @ ./test.jl:56             <<
in expression starting at ./test.jl:56
```

I suggest that fixing the above requires coordinated input from both `Internal` and `Exposed`. The former would need to capture the end-user backtrace so as to forward it to the latter:
```diff
struct CallBackError <: Exception
    err
+   bt::Base.ExceptionStack
end
  ...
- rethrow(CallBackError(e))
+ rethrow(CallbackError(e, current_exceptions()))
```
While the latter would accept this payload and only dismiss stackframes in-between:
```diff
- rethrow(NiceReport(e.user_err))
+ rethrow(NiceReport(e.user_err); stitch = e.bt)
```

Perhaps situations A and B cover most (if not all ?) situations where the end user ends up seeing irrelevant stack frames.
A stack frame is irrelevant if it points into code with no bug that the user is not responsible for.
That "no bug" guarantee can only be held by lib code authors, using their own means to prove it (like the pattern `e isa Expected || rethrow(e)` above) and taking responsibility for it.

What do you think?

**Comment by BioTurboNick** [comment]
A: This is in theory a standard best practice for exceptions for any language. However, the added boilerplate at public APIs just to make the stacktrace nicer could be resisted.

B: Seems like a reasonable feature to have.

Having tools to help package authors reduce stacktraces is nice, but that requires an ecosystem-wide effort to adopt and education to help new packages adopt it. It could help long-term, but IMO users really need something they can control when package authors don't, or can't, do much to trim traces.

**Comment by iago-lito** [comment]
I see @BioTurboNick. IIUC you are arguing that not only (1) tooling to help lib authors reduce the stack for their users is desired, but also (2) tooling to help lib users reduce the stack themselves, possibly default-configured by Julia.

I agree that these are separate concerns. And that (from it's title) this issue is about (2) not (1).

- (1) is good/important to have. Today we do not even have (1). If (1) becomes spread and used at least by `Base` and and a few major libs out there like @ChrisRackauckas' `DiffEq`'s suite, then I expect that the urge for (2) might become weaker and weaker as good practices distill. Should I open a separate issue/request about possible `(re)throw(e; afresh = true)` and/or `(re)throw(e; stitch = bt)`? (note that the former is arguably a special case of the latter)

- I wouldn't be surprised that (2) be a fantasy. Users cannot ultimately decide which section of a stack is irrelevant when it's deep into code they have no knowledge of. I suppose we could craft heuristics and let users in charge of fine-tuning them but I expect everlasting quirks with such an approach because I don't think it ultimately makes sense.

**Comment by stevengj** [comment]
I still feel like a better default in 99% of cases would be abbreviated stack traces ala https://github.com/JuliaLang/julia/issues/40138#issuecomment-853401449 or similar ideas.  There can be a function, e.g. `fullstacktrace()` in InteractiveUtils (loaded by the REPL etc.) to print the full stack trace of the most recent exception upon request, of course.

**Comment by BioTurboNick** [comment]
> * (1) is good/important to have. Today we do not even have (1). If (1) becomes spread and used at least by `Base` and and a few major libs out there like [@ChrisRackauckas](https://github.com/ChrisRackauckas)' `DiffEq`'s suite, then I expect that the urge for (2) might become weaker and weaker as good practices distill.

Tbh, don't hold your breath about the pace of change here.

> * Should I open a separate issue/request about possible `(re)throw(e; afresh = true)` and/or `(re)throw(e; stitch = bt)`? (note that the former is arguably a special case of the latter)

`CallbackError` as a separate issue, or if you feel so inclined, just go ahead and propose a PR to that effect. Could be a debate about it, but seems useful to have a code representation to work off of, with demonstration of its utility in e.g., `map`.

> * I wouldn't be surprised that (2) be a fantasy. Users cannot ultimately decide which section of a stack is irrelevant when it's deep into code they have no knowledge of. I suppose we could craft heuristics and let users in charge of fine-tuning them but I expect everlasting quirks with such an approach because I don't think it ultimately makes sense.

The latter has already been proven to work for years now, for those who know about my package AbbreviatedStackTraces.jl and don't mind the recompilation overhead. If you have an example where it wouldn't work for you, I would be happy to hear what it is.

**Comment by BioTurboNick** [comment]
> I still feel like a better default in 99% of cases would be abbreviated stack traces ala [#40138 (comment)](https://github.com/JuliaLang/julia/issues/40138#issuecomment-853401449) or similar ideas. There can be a function, e.g. `fullstacktrace()` in InteractiveUtils (loaded by the REPL etc.) to print the full stack trace of the most recent exception upon request, of course.

And the last N full stacktraces could be saved to a log, even.
